### DISCUSSION TITLE
Review workflow secrets security

### DISCUSSION THREAD

**Discussion by josecelano** [Nautilus-Cyberneering/git-queue#147] [Ideas]
@da2ce7 and I were discussing workflow secrets. We have to review them in order to avoid someone stealing the secrets.

I'm going to start the discussion with some useful links to read before going deep into the details.

### LINKS

- GitHub docs: [Implementing least privilege for secrets in GitHub Actions](https://github.blog/2021-04-13-implementing-least-privilege-for-secrets-in-github-actions/)
- GitHub docs: [Environment secrets](https://docs.github.com/en/actions/deployment/targeting-different-environments/using-environments-for-deployment#environment-secrets).
- [Using CODEOWNERS to monitor changes](https://docs.github.com/en/actions/security-guides/security-hardening-for-github-actions#using-codeowners-to-monitor-changes). You can use the CODEOWNERS feature to control how changes are made to your workflow files
- GitHub docs: [Deployment branches](https://docs.github.com/en/actions/deployment/targeting-different-environments/using-environments-for-deployment#deployment-branches).
- GitHub Community Forum: [Limit secrets to specific branches](https://github.community/t/limit-secrets-to-specific-branches/18240)
- [branch-based-secrets GitHub Action](https://github.com/noliran/branch-based-secrets). Dynamically use different secrets depending on the current branch the action is currently running on.

- [Keeping your GitHub Actions and workflows secure Part 1: Preventing pwn requests](https://securitylab.github.com/research/github-actions-preventing-pwn-requests/).
- [Github Actions and the threat of malicious pull requests](https://nathandavison.com/blog/github-actions-and-the-threat-of-malicious-pull-requests).
- [Using Environment Protection Rules to Secure Secrets When Building External Forks with pull_request_target](https://dev.to/petrsvihlik/using-environment-protection-rules-to-secure-secrets-when-building-external-forks-with-pullrequesttarget-hci)
- [GitHub Actions & Security: Best practices](https://devopsjournal.io/blog/2021/02/06/GitHub-Actions)

- Staroverflow question: [How to Define Github Branch Specific Secret or an Alternative?](https://stackoverflow.com/questions/59097530/how-to-define-github-branch-specific-secret-or-an-alternative).

- Blog post: [Stealing arbitrary GitHub Actions secrets](https://blog.teddykatz.com/2021/03/17/github-actions-write-access.html).

**Comment by josecelano** [comment]
List of secrets we are using:

Repository secrets:

- GPG_PRIVATE_KEY: a GPG key used only for testing purposes to sign commits on the [test](https://github.com/Nautilus-Cyberneering/git-queue/blob/main/.github/workflows/test.yml) workflow.
- PASSPHRASE: passphrase for the previous GPG key.

Other secrets:

- GITHUB_TOKEN: used by `megalinter` and another action to create releases: `ncipollo/release-action@v1`

**Comment by josecelano** [comment]
I have changed the organization's default workflow permissions:

![image](https://user-images.githubusercontent.com/58816/164249361-acc324a1-e90e-4ff5-8a09-534306258f4f.png)

I think it will help us to have a more restricted approach and also to be more explicit in defining permissions for the workflows.

One of the [test projects](https://github.com/Nautilus-Cyberneering/library-aaa)' workflow has failed:

Previous default permissions:

![image](https://user-images.githubusercontent.com/58816/164250391-aa2e93ab-08c8-4453-be12-79c216da9bf9.png)

New default permissions:

![image](https://user-images.githubusercontent.com/58816/164250241-1ae7b5ef-fef4-4133-8884-a5359f24a3a6.png)

I suppose we need to add the `contents: write` permission.

[List of permissions](https://docs.github.com/en/actions/security-guides/automatic-token-authentication#permissions-for-the-github_token).

**Reply by josecelano** [reply]
The [problem was fixed](https://github.com/Nautilus-Cyberneering/library-aaa/pull/3) by adding the "contents: write" permission to the workflow job.

```yml
update:
  runs-on: ubuntu-latest
  permissions:
    issues: write
    contents: write
  steps:
    - uses: actions/checkout@v2
```

**Reply by josecelano** [reply]
The other test project ([library-consumer](https://github.com/Nautilus-Cyberneering/library-consumer)) has also [failed](https://github.com/Nautilus-Cyberneering/library-consumer/issues/25), probably for the same reason.

**Comment by da2ce7** [comment]
https://securitylab.github.com/research/github-actions-preventing-pwn-requests/
https://securitylab.github.com/research/github-actions-untrusted-input/
https://securitylab.github.com/research/github-actions-building-blocks/

**Reply by josecelano** [reply]
I have installed the GitHub code scanning as described in the second part of the article and [some security errors have been found](https://github.com/Nautilus-Cyberneering/git-queue/pull/160/checks?check_run_id=6128888722).

**Comment by josecelano** [comment]
This is in general a list of nice-to-have things about security: https://bestpractices.coreinfrastructure.org/en/projects?q=git-queue

**Comment by josecelano** [comment]
@da2ce7 regarding not trusting third-party actions like for example the one used by MegaLinter, right now with the most restricted workflow permissions the MegaLinter workflow GITHUB_TOKEN would have these permissions:

- [Contents: read](https://docs.github.com/en/rest/overview/permissions-required-for-github-apps#permission-on-contents)
- [Metadata: read](https://docs.github.com/en/rest/overview/permissions-required-for-github-apps#metadata-permissions)

And it seems the workflow has only those permissions even if [you execute it having write permissions with a local branch](https://github.com/Nautilus-Cyberneering/git-queue/runs/6128860491?check_suite_focus=true#step:1:13).

![image](https://user-images.githubusercontent.com/58816/164738619-066ef008-46ce-44ba-8e26-088b624591f0.png)

**Comment by josecelano** [comment]
Creating a PR from a fork shows this warning:

![image](https://user-images.githubusercontent.com/58816/165077584-2118178c-7fda-4b7e-ad78-9c872bca9e63.png)

[More info](https://docs.github.com/en/pull-requests/collaborating-with-pull-requests/working-with-forks/allowing-changes-to-a-pull-request-branch-created-from-a-fork).
